server.js

```diff
@@ -6801,6 +6801,119 @@ cache((data, match, sendBadge, request) => {
   });
 }));
 
+// nsp for npm packages
+camp.route(/^\/nsp\/npm\/(@([^\/]+)\/)?[^\/]+(\/[^\/]+)?\.(svg|png|gif|jpg|json)$/, cache((data, match, sendBadge, request) => {
+  // A: /nsp/npm/:package.:format
+  // B: /nsp/npm/:package/:version.:format
+  // C: /nsp/npm/@:scope/:package.:format
+  // D: /nsp/npm/@:scope/:package/:version.:format
+  const pieces = match.input.split('/');
+  let badgeData = getBadgeData('nsp', data);
+  let format = '';
+  let nspRequestOptions = {
+    method: 'POST',
+    body: {
+      package: {}
+    },
+    json: true
+  };
+
+  function getNspResults () {
+    request('https://api.nodesecurity.io/check', nspRequestOptions, (error, response, body) => {
+      if (error !== null || typeof body !== 'object' || body === null) {
+        badgeData.text[1] = 'invalid';
+        badgeData.colorscheme = 'red';
+      } else if (body.length !== 0) {
+        badgeData.text[1] = `${body.length} vulnerabilities`;
+        badgeData.colorscheme = 'red';
+      } else {
+        badgeData.text[1] = 'no known vulnerabilities';
+        badgeData.colorscheme = 'brightgreen';
+      }
+
+      sendBadge(format, badgeData);
+    });
+  }
+
+  function getNpmVersionThenNspResults (packageName = '') {
+    // nsp doesn't properly detect the package version in POST requests so this function gets it for us
+    // https://github.com/npm/registry/blob/master/docs/REGISTRY-API.md#getpackageversion
+    let npmURL;
+    const npmRequestOptions = {
+      headers: {
+        Accept: '*/*'
+      },
+      json: true
+    };
+
+    try {
+      if (packageName[0] === '@') {
+        // Using 'latest' would save bandwidth, but it is currently not supported for scoped packages
+        npmURL = `http://registry.npmjs.org/@${encodeURIComponent(packageName.slice(1))}`;
+      } else {
+        npmURL = `http://registry.npmjs.org/${encodeURIComponent(packageName)}/latest`;
+      }
+    } catch (e) {
+      badgeData.text[1] = 'invalid';
+      badgeData.colorscheme = 'red';
+
+      sendBadge(format, badgeData);
+      return;
+    }
+
+    request(npmURL, npmRequestOptions, (error, response, body) => {
+      if (response !== null && response.statusCode === 404) {
+        // NOTE: in POST requests nsp does not distinguish between 'package not found' and 'no known vulnerabilities'. To keep consistency in the use case where a version is provided (which skips `getNpmVersionThenNspResults()` altogether) we'll say 'no known vulnerabilities' since it is technically true in both cases
+        badgeData.text[1] = 'no known vulnerabilities';
+
+        sendBadge(format, badgeData);
+      } else if (error !== null || typeof body !== 'object' || body === null) {
+        badgeData.text[1] = 'invalid';
+        badgeData.colorscheme = 'red';
+
+        sendBadge(format, badgeData);
+      } else {
+        nspRequestOptions.body.package.version = body.version;
```

Comment:
Perhaps this would be easier to reason about if it passed the version into `getNspResults` as a parameter.